File: src/adapters/driver/http/handlers/allStreamsPlaylistHandler_test.go
package handlers

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"Theatrum/domain/models"
	"Theatrum/domain/services"
)

// handlerMockStorage implements repositories.StoragePort for testing.
type handlerMockStorage struct {
	searchFilesFn func(pattern string, extensions []string) ([]string, []map[string]string, error)
}

func (m *handlerMockStorage) ReadFile(path string) ([]byte, error)     { return nil, nil }
func (m *handlerMockStorage) WriteFile(path string, data []byte) error { return nil }
func (m *handlerMockStorage) DeleteFile(path string) error             { return nil }
func (m *handlerMockStorage) ListFiles(pattern string) ([]string, error) {
	return nil, nil
}
func (m *handlerMockStorage) GetFileSize(path string) (int64, error) { return 0, nil }
func (m *handlerMockStorage) SearchFiles(pattern string, extensions []string) ([]string, []map[string]string, error) {
	if m.searchFilesFn != nil {
		return m.searchFilesFn(pattern, extensions)
	}
	return nil, nil, nil
}

func TestAllStreamsPlaylistHandler_ServeHTTP(t *testing.T) {
	t.Run("builds playlist with correct content-type", func(t *testing.T) {
		app := &models.Application{
			PublicPath: "http://localhost:8080",
			AllStreamsPlaylist: models.AllStreamsPlaylist{
				Enabled: true,
				Path:    "/all.m3u8",
			},
		}
		channels := map[string]models.Stream{
			"/vod/{name}": {
				Type: models.StreamTypeVideoEncoded,
				Path: "videos/{name}",
			},
		}
		storage := &handlerMockStorage{
			searchFilesFn: func(pattern string, extensions []string) ([]string, []map[string]string, error) {
				return []string{"videos/test/master.m3u8"}, []map[string]string{{"name": "test"}}, nil
			},
		}
		templateSvc := services.NewPathTemplateService()
		appSvc := services.NewApplicationService(app, &models.Server{}, &channels, storage, templateSvc)
		streamSvc := services.NewStreamService(templateSvc)

		handler := NewAllStreamsPlaylistHandler(appSvc, streamSvc)

		req := httptest.NewRequest("GET", "/all.m3u8", nil)
		w := httptest.NewRecorder()

		handler.ServeHTTP(w, req)

		if w.Code != http.StatusOK {
			t.Errorf("expected status 200, got %d", w.Code)
		}
		if ct := w.Header().Get("Content-Type"); ct != "application/x-mpegURL" {
			t.Errorf("expected Content-Type application/x-mpegURL, got %q", ct)
		}
		body := w.Body.String()
		if !strings.Contains(body, "#EXTM3U") {
			t.Error("expected M3U8 header in response body")
		}
	})

	t.Run("error handling when disabled", func(t *testing.T) {
		app := &models.Application{
			AllStreamsPlaylist: models.AllStreamsPlaylist{Enabled: false},
		}
		channels := map[string]models.Stream{}
		storage := &handlerMockStorage{}
		templateSvc := services.NewPathTemplateService()
		appSvc := services.NewApplicationService(app, &models.Server{}, &channels, storage, templateSvc)
		streamSvc := services.NewStreamService(templateSvc)

		handler := NewAllStreamsPlaylistHandler(appSvc, streamSvc)

		req := httptest.NewRequest("GET", "/all.m3u8", nil)
		w := httptest.NewRecorder()

		handler.ServeHTTP(w, req)

		if w.Code != http.StatusInternalServerError {
			t.Errorf("expected status 500, got %d", w.Code)
		}
	})

	t.Run("error handling when storage fails", func(t *testing.T) {
		app := &models.Application{
			PublicPath: "http://localhost:8080",
			AllStreamsPlaylist: models.AllStreamsPlaylist{
				Enabled: true,
				Path:    "/all.m3u8",
			},
		}
		channels := map[string]models.Stream{
			"/vod/{name}": {
				Type: models.StreamTypeVideoEncoded,
				Path: "videos/{name}",
			},
		}
		storage := &handlerMockStorage{
			searchFilesFn: func(pattern string, extensions []string) ([]string, []map[string]string, error) {
				return nil, nil, fmt.Errorf("storage error")
			},
		}
		templateSvc := services.NewPathTemplateService()
		appSvc := services.NewApplicationService(app, &models.Server{}, &channels, storage, templateSvc)
		streamSvc := services.NewStreamService(templateSvc)

		handler := NewAllStreamsPlaylistHandler(appSvc, streamSvc)

		req := httptest.NewRequest("GET", "/all.m3u8", nil)
		w := httptest.NewRecorder()

		handler.ServeHTTP(w, req)

		// Should still return 200 with partial playlist (errors are logged, not fatal)
		if w.Code != http.StatusOK {
			t.Errorf("expected status 200, got %d", w.Code)
		}
	})
}

File: src/test/java/com/fjourdren/theatrum/infrastructure/adapter/in/web/handlers/AllStreamsPlaylistHandlerTest.java
package com.fjourdren.theatrum.infrastructure.adapter.in.web.handlers;

import com.fjourdren.theatrum.application.port.out.StoragePort;
import com.fjourdren.theatrum.domain.model.AllStreamsPlaylist;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.Application;
import com.fjourdren.theatrum.domain.model.FileMatch;
import com.fjourdren.theatrum.domain.model.LoadedConfiguration;
import com.fjourdren.theatrum.domain.model.Server;
import com.fjourdren.theatrum.domain.model.Stream;
import com.fjourdren.theatrum.domain.model.StreamType;
import com.fjourdren.theatrum.domain.service.ApplicationService;
import com.fjourdren.theatrum.domain.service.PathTemplateService;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.mock.web.MockHttpServletResponse;

import java.io.IOException;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

/** Port of Go's {@code allStreamsPlaylistHandler_test.go}. */
@ExtendWith(MockitoExtension.class)
class AllStreamsPlaylistHandlerTest {

    private static final AppPaths PATHS = new AppPaths(Path.of("data"), Path.of("frontend"));

    @Mock
    private StoragePort storage;

    private final MockHttpServletResponse response = new MockHttpServletResponse();

    private AllStreamsPlaylistHandler handler(Application app, Map<String, Stream> channels) {
        var applicationService = new ApplicationService(
                new LoadedConfiguration(app, new Server(8080, 1935, null), channels),
                storage, new PathTemplateService(), PATHS);
        return new AllStreamsPlaylistHandler(applicationService);
    }

    private static Map<String, Stream> vodChannel() {
        var channels = new LinkedHashMap<String, Stream>();
        channels.put("/vod/{name}", Stream.builder().type(StreamType.VIDEO_ENCODED).path("videos/{name}").build());
        return channels;
    }

    @Test
    void buildsThePlaylistWithTheCorrectContentType() throws IOException {
        when(storage.searchFiles(anyString(), any())).thenReturn(
                List.of(new FileMatch(Path.of("videos/test/master.m3u8"), Map.of("name", "test"))));

        handler(new Application("http://localhost:8080", new AllStreamsPlaylist(true, "/all.m3u8")), vodChannel())
                .handle(response);

        assertThat(response.getStatus()).isEqualTo(200);
        assertThat(response.getContentType()).isEqualTo("application/x-mpegURL");
        assertThat(response.getContentAsString()).contains("#EXTM3U");
        assertThat(response.getContentAsString()).contains("http://localhost:8080/vod/test/master.m3u8");
    }

    @Test
    void returns500WhenTheFeatureIsDisabled() throws IOException {
        handler(new Application("", new AllStreamsPlaylist(false, "")), new LinkedHashMap<>())
                .handle(response);

        assertThat(response.getStatus()).isEqualTo(500);
    }

    /** Storage failures are logged inside the service, so a partial playlist is still returned. */
    @Test
    void returns200WithAPartialPlaylistWhenStorageFails() throws IOException {
        when(storage.searchFiles(anyString(), any())).thenThrow(new IOException("storage error"));

        handler(new Application("http://localhost:8080", new AllStreamsPlaylist(true, "/all.m3u8")), vodChannel())
                .handle(response);

        assertThat(response.getStatus()).isEqualTo(200);
        assertThat(response.getContentAsString()).contains("#EXTM3U");
    }
}
